Take advantage of Jersey functionality

blootsvoets opened the pull request:
- Use Jersey exception mappers to handle exceptions, instead of repetitive try/catch blocks. Also document 404 error states.
- Use dependency injection instead of web listeners.
- Use Jersey filters instead of servlet filters.
- Use annotations to handle security

blootsvoets committed:
Use ExceptionMapper instead of repeated try/catch blocks

blootsvoets committed:
Fixed some annotations and added separate IOExceptionMapper

blootsvoets committed:
Using injects instead of listeners

blootsvoets committed:
Merge branch 'dev' into exception-mapping

blootsvoets committed:
Use annotations to manage endpoint security

blootsvoets committed:
Update radar-auth and documentation

yatharthranjan commented on src/main/java/org/radarcns/listener/managementportal/ManagementPortalClient.java:
```diff
@@ -111,8 +112,12 @@ private static Duration parseDuration(String duration, Duration defaultValue) {
+        this.token = tokenDetails.getAccessToken();
+    }
+
+    private synchronized String getToken() {
```
Why does this need to be synchronized? Is it for when a token is being updated by the above function ?

blootsvoets replied:
exactly

blootsvoets committed:
Mostly functional DI

Still TODO: add back Swagger and fix default headers when no data
is available.

blootsvoets committed:
Fixed gradle check and added todo's

blootsvoets committed:
Merge AVRO_BINARY and APPLICATION_JSON paths

Plus:
- add @Alphanumeric annotation to check that fields are valid.
- Fixed some integration tests.

Still missing: swagger support

blootsvoets committed:
Removed unused Param class and moved security -> auth

blootsvoets committed:
Removed ResponseHandler and documented exception mappers

blootsvoets committed:
Code style fixes

nivemaham commented on src/main/java/org/radarcns/listener/MongoFactory.java:
```diff
@@ -0,0 +1,75 @@
+
+/**
+ * pon the web application initialisation, this Context Listener creates a MongoDb client that can
+ * be reused by each call. A Mongo Client should be seen like a Thread Pool.
```
The javadocs should be updated as well+-

nivemaham commented on src/main/java/org/radarcns/listener/managementportal/ManagementPortalClientFactory.java:
```diff
@@ -0,0 +1,83 @@
+            oauthClient = new OAuth2Client.Builder()
+                    .endpoint(url)
+                    .credentials(clientId, clientSecret)
+                    .httpClient(httpClient)
```
We should also initialize scopes of the client.

nivemaham commented on src/main/java/org/radarcns/webapp/resource/SensorEndPoint.java:
```diff
@@ -0,0 +1,246 @@
+                    + "time-frame resolutions. The response is formatted in JSON.")
+    @ApiResponse(responseCode = "500", description = "An error occurs while executing, in the body "
+            + "there is a message.avsc object with more details.")
+    @ApiResponse(responseCode = "204", description =
```
We should update the javadocs to say 404 too

blootsvoets replied:
Done

nivemaham commented on src/main/java/org/radarcns/auth/PermissionFilter.java:
```diff
@@ -0,0 +1,54 @@
+                requestContext.getUriInfo().getPath(), message);
+        Response.ResponseBuilder builder = Response.status(Status.FORBIDDEN);
+        if (requestContext.getMediaType() == null
+                || requestContext.getMediaType().isCompatible(APPLICATION_JSON_TYPE)) {
```
Is `AVRO-BINARY` compatible with `JSON` type? Would this also work when a request is send just with `avro` as the accepted response type?

blootsvoets replied:
No, because `StatusMessage` is not a `SpecificRecord`, it can't be encoded as `avro/binary`

blootsvoets committed:
javadoc and style fixes

nivemaham commented on src/main/java/org/radarcns/listener/managementportal/ManagementPortalClientFactory.java:
```diff
@@ -0,0 +1,76 @@
+
+    @Override
+    public void dispose(ManagementPortalClient instance) {
+        // no disposal needed
```
Don't we have to clear up the oauth2-client, at least the reference of httpClient to it ?

blootsvoets replied:
The OkHttpClient will be disposed of by the `HttpClientFactory`. `OAuth2Client` does not have any disposal/close functionality, nor does the `ManagementPortalClient`.

blootsvoets committed:
Javadoc fixes

nivemaham commented on src/main/java/org/radarcns/webapp/filter/AuthenticationFilter.java:
```diff
@@ -0,0 +1,92 @@
+            logger.warn("[401] {}: No token bearer header provided in the request",
+                    requestContext.getUriInfo().getPath());
+            requestContext.abortWith(
+                    Response.status(Status.UNAUTHORIZED)
```
would it be possible to add a message as well in the response, to differentiate between e.g. unauthorized token (invalid scope) and token not available which would have 401?

blootsvoets replied:
Done, it now has the `token_missing` and `token_invalid` errors with a message.

blootsvoets committed:
Documentation fixes

nivemaham commented on src/main/java/org/radarcns/webapp/filter/CorsFilter.java:
```diff
@@ -0,0 +1,24 @@
+/** Standard CORS filter, applied to all responses. */
+@Provider
+@Priority(1000)
+public class CorsFilter implements ContainerResponseFilter {
```
Doesn't this has to be registered to the context as well? Or is it registered by default?

blootsvoets replied:
Filters with an `@Provider` tag, in a directory that is scanned by `RadarApplication`, are automatically added.

blootsvoets committed:
Added status message on unauthorized

blootsvoets committed:
CORS higher in priority

nivemaham commented on README.md:
```diff
@@ -19,7 +19,7 @@ To deploy the war do:
 By default, log messages are redirected to the `STDOUT`.
 
-The api documentation is located at `<your-server-address>:<port>/radar/api/openapi.json`
+The api documentation is located at `<your-server-address>:<port>/api/openapi.json`
```
I think we should also change the url in restapi-details.yml in integrationTest to match this.

blootsvoets replied:
yup, already done

nivemaham replied:
Very nice and neat work. I will start merging this branch with the `removing-device-catalog` branch which also has significant amount of refactoring and will result in large merge conflict. Once approved we will update from dev.

blootsvoets committed:
Travis build update

blootsvoets committed:
Removed 204 status codes

blootsvoets committed:
Javadoc additions

blootsvoets committed:
Javadoc fix

yatharthranjan commented on src/main/java/org/radarcns/auth/RadarSecurityContext.java:
```diff
@@ -0,0 +1,53 @@
+     *         {@code false} otherwise
+     */
+    @Override
+    public boolean isUserInRole(String role) {
```
Can you please explain this a bit more?

blootsvoets replied:
It is just an extension of the standard Jersey SecurityContext. It allows annotations like
`@RolesAllowed(":SYS_ADMIN")`, without further detailed checking of the `RadarToken`. See https://jersey.github.io/documentation/latest/security.html

yatharthranjan replied:
Thanks. Did not know that !

blootsvoets committed:
Added example to RadarSecurityContext#isUserInRole

yatharthranjan commented on src/main/java/org/radarcns/webapp/resource/ManagementPortalEndPoint.java:
```diff
@@ -0,0 +1,101 @@
+ */
+@Provider
+@Path("/mp")
+public class ManagementPortalEndPoint {
```
Can we rename this to SubjectEndPoint and remove the old one because all subject level information has to be taken from the Management Portal ?

blootsvoets replied:
Yes, in the next PR please.

yatharthranjan commented on src/main/java/org/radarcns/webapp/filter/AuthorizationFeature.java:
```diff
@@ -0,0 +1,31 @@
+            context.register(PermissionOnProjectFilter.class, 2050);
+        }
+        if (method.isAnnotationPresent(NeedsPermissionOnSubject.class)) {
+            context.register(PermissionOnSubjectFilter.class, 2100);
```
How was the priority decided here?

blootsvoets replied:
In the end, it is just an in-app ordering
1. Cors (500)
2. Authentication (1000)
3. Authorization (permission only) (2000)
4. Authorization (permission + project) (2050)
5. Authorization (permission + project) (2100)

It ensures the order, and leaves plenty of room for other filters to be added in between.

yatharthranjan replied:
Okay great. Typo there (permission + subject) but got it! Thanks

yatharthranjan reviewed (approved):
Very nice work! Looks good to me.
